2.0/m_apacheauth.cpp

```diff
@@ -0,0 +1,159 @@
+		authfile = conf->getString("authfile");
+		killreason = conf->getString("killreason");
+		allowpattern = conf->getString("allowpattern");
+		verbose = conf->getBool("verbose");
+		LoadAuthFile();
+	}
+
+	bool SplitRecord(const std::string &record, std::string &record_algorithm, std::string &record_salt, std::string &record_hash)
+	{
+		size_t i[3];
+
+		for (size_t pos = 0, n = 0; n < 3; ++n) {
+			i[n] = pos = record.find('$', pos);
+			if (pos == std::string::npos) return false;
+			++pos;
+		}
+		record_algorithm = record.substr(i[0] + 1, i[1]-i[0] - 1);
+		record_salt = record.substr(i[1] + 1, i[2]-i[1] - 1);
+		record_hash = record.substr(i[2] + 1);
+		return true;
+	}
+
+	void LoadAuthFile()
+	{
+		int line=0;
+		std::ifstream f;
+
+		logins.clear();
+		ServerInstance->SNO->WriteGlobalSno('a', "Loading auth file %s",authfile.c_str());
+		f.open(authfile.c_str(), std::ifstream::in);
+		if (!f.good()) {
+			ServerInstance->SNO->WriteGlobalSno('a', "Auth file failed to open, no connections will be allowed");
+			return;
+		}
+		std::string s;
+		while (std::getline(f, s)) {
+			++line;
+			size_t pos = s.find(':');
+			if (pos == std::string::npos) {
+				ServerInstance->SNO->WriteGlobalSno('a', "Syntax error at line %d", line);
+				continue;
+			}
+			std::string login = s.substr(0, pos);
+			std::string hashed = s.substr(pos + 1);
+			if (verbose) {
+				ServerInstance->SNO->WriteGlobalSno('a', "Found login %s, hash %s", login.c_str(), hashed.c_str());
+			}
+			if (logins.find(login) != logins.end()) {
+				ServerInstance->SNO->WriteGlobalSno ('a', "Warning: ignoring duplicate login: %s", login.c_str());
+			}
+			else {
+				HashedPassword p;
+				p.hashed = hashed;
+				if (!SplitRecord(hashed, p.algorithm, p.salt, p.hash)) {
+					ServerInstance->SNO->WriteGlobalSno ('a', "Warning: ignoring malformed line: %s", hashed.c_str());
+				}
+				else if (p.algorithm != "1") { // MD5
+					ServerInstance->SNO->WriteGlobalSno ('a', "Warning: ignoring unsupported algorithm: %s", p.algorithm.c_str());
+				}
+				else {
+					logins.insert(std::make_pair(login, p));
+				}
+			}
+		}
+		ServerInstance->SNO->WriteGlobalSno ('a', "Done C++ loading auth file, %u users", (unsigned)logins.size());
+		f.close();
+	}
+
+	ModResult OnUserRegister(LocalUser* user)
+	{
+		// Note this is their initial (unresolved) connect block
+		ConfigTag* tag = user->MyClass->config;
+		if (!tag->getBool("useapacheauth", true))
+			return MOD_RES_PASSTHRU;
+
+		if (!allowpattern.empty() && InspIRCd::Match(user->ident,allowpattern))
+			return MOD_RES_PASSTHRU;
+
+		//ServerInstance->SNO->WriteGlobalSno('a', "Got request to auth %s with hash %s", user->nick.c_str(), user->password.c_str());
+		std::map<std::string, HashedPassword>::const_iterator i = logins.find(user->ident);
+		if (i == logins.end()) {
+			ServerInstance->SNO->WriteGlobalSno('a', "Forbiding connection from %s!%s@%s (login not found)",
```

Comment:
typo in message, I'd suggest "Denying connection" instead perhaps.

while you're here, can I suggest that you remove the commented out lines?